### PR TITLE
Support new constants in JDK 9 and 10

### PR DESCRIPTION
**Pull request by retronym**

### CODE REVIEW

**Comment by typesafe-tools** [comment]
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?sbt=1.1.x&zinc=pull/522/head&io=1.1.x&librarymanagement=1.1.x&util=1.1.x&website=1.1.x&deploy_to_bintray=false&run_sbt_community=false&verbose_dbuild=false&community_verbose=false">this link.</A>

**Comment by retronym** [comment]
I still need to test this.

Here's what happens at the moment:

```
[info] Done updating.
[info] Done compiling.
[error] java.lang.RuntimeException: Unknown constant: 19
[error] 	at scala.sys.package$.error(package.scala:27)
[error] 	at sbt.internal.inc.classfile.Parser$.getConstant(Parser.scala:163)
[error] 	at sbt.internal.inc.classfile.Parser$.parse$1(Parser.scala:129)
[error] 	at sbt.internal.inc.classfile.Parser$.sbt$internal$inc$classfile$Parser$$parseConstantPool(Parser.scala:134)
[error] 	at sbt.internal.inc.classfile.Parser$$anon$1.<init>(Parser.scala:42)
[error] 	at sbt.internal.inc.classfile.Parser$.parseImpl(Parser.scala:38)
[error] 	at sbt.internal.inc.classfile.Parser$.parse(Parser.scala:33)
[error] 	at sbt.internal.inc.classfile.Parser$.$anonfun$apply$1(Parser.scala:27)
[error] 	at sbt.io.Using.apply(Using.scala:22)
[error] 	at sbt.internal.inc.classfile.Parser$.apply(Parser.scala:27)
```

Compiling https://github.com/retronym/javafileman/tree/zinc-issue, which contains `module-info.java` files, with SBT 1.1.2 and Java 9.

**Comment by retronym** [comment]
I've tested this locally and compile is working as expected. The second commit avoids a warning in the part of `AnalyzingJavaCompiler` that uses Java reflection to establish the class API. We should squash these on merge.

File: internal/zinc-classfile/src/main/scala/sbt/internal/inc/classfile/ClassFile.scala [hunk 1]
```diff
@@ -106,5 +106,8 @@ private[sbt] object Constants {
   final val ConstantMethodHandle = 15
   final val ConstantMethodType = 16
   final val ConstantInvokeDynamic = 18
+  final val ConstantModule = 19
+  final val ConstantPackage = 20
+  final val ConstantDynammic = 17 // http://openjdk.java.net/jeps/309
```
Comment (jvican) [inline internal/zinc-classfile/src/main/scala/sbt/internal/inc/classfile/ClassFile.scala]: Typo: `ConstantDynamic` :smile:

**Comment by retronym** [comment]
Flaky test?

```
[info] zinc
[info] - should cache jar generation *** FAILED ***
[info]   24 was not less than 20.0 Cache jar didn't work: 24 is >= than 20% of 100. (CachedHashingSpec.scala:61)
```

**Review by dwijnand — APPROVED** [review]

**Review by jvican — APPROVED** [review]

**Pull request merged**